impl/src/expand.rs

```diff
@@ -377,6 +547,132 @@ fn impl_enum(input: Enum) -> TokenStream {
     }
 }
 
+#[cfg_attr(test, derive(Debug))]
+#[derive(Clone)]
+struct GenericUsageVisitor {
+    generics: std::collections::HashMap<proc_macro2::Ident, bool>,
+}
+
+impl GenericUsageVisitor {
+    pub fn new<TPairs>(generics: TPairs) -> Self
+    where
+        TPairs: IntoIterator<Item = (syn::Ident, bool)>,
+    {
+        Self {
+            generics: generics.into_iter().collect(),
+        }
+    }
+
+    pub fn new_unmarked<TIdents>(generics: TIdents) -> Self
+    where
+        TIdents: IntoIterator<Item = syn::Ident>,
+    {
+        Self::new(generics.into_iter().map(|ident| (ident, false)))
+    }
+
+    pub fn iter_marked(&self) -> impl Iterator<Item = &proc_macro2::Ident> {
+        self.generics.iter().filter(|(_k, v)| **v).map(|(k, _v)| k)
+    }
+
+    #[allow(dead_code)]
+    pub fn is_marked<'a, T: PartialEq<&'a proc_macro2::Ident> + 'a>(&'a self, item: T) -> bool {
+        self.iter_marked().any(|ident| item == ident)
+    }
+
+    pub fn mark_from<'a, TMarkedSource: IntoIterator<Item = &'a proc_macro2::Ident> + 'a>(
+        &'a mut self,
```

Comment:
The `'a` lifetime requirements in this signature are never used.

```diff
- pub fn mark_from<'a, TMarkedSource: IntoIterator<Item = &'a proc_macro2::Ident> + 'a>(
-     &'a mut self,
+ pub fn mark_from<'a, TMarkedSource: IntoIterator<Item = &'a proc_macro2::Ident>>(
+     &mut self,
```

Reply:
Hum. I thought if I had one explicit (the one used in `IntoIterator::Item`) they were supposed to all be explicit. Neat.